### PR TITLE
Add drops to shear events

### LINKED ISSUE #5672: Add PlayerShearEntityEvent#getDrops method
### Is your feature request related to a problem?

Currently there is no easy way to get the drops of the `PlayerShearEntityEvent`

### Describe the solution you'd like.

Some way to get the drops of shearing.
Taking into account the enchantments used.
Allowing the user to set the drops as well.
Example: Shearing a sheep and instead of it dropping 3 white wool it drops 3 random colors

### Describe alternatives you've considered.

Right now you'd have to listenten to multiple events to figure out what items are dropped

### Agreements

- [X] I have searched for and ensured there isn't already an open issue regarding this.
- [X] I have ensured the feature I'm requesting isn't already in the latest supported Paper build.

### Other

_No response_

### PR DESCRIPTION
**Pull request by Machine-Maker**
Closes #5672

### CODE REVIEW

File: Spigot-API-Patches/0304-Add-drops-to-shear-events.patch [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+     }
+ 
+     /**
+@@ -64,4 +72,25 @@ public class BlockShearEntityEvent extends BlockEvent implements Cancellable {
+     public static HandlerList getHandlerList() {
+         return handlers;
+     }
++    // Paper start
++    /**
++     * Get an immutable list of drops for this shearing.
```
Comment (Proximyst) [inline Spigot-API-Patches/0304-Add-drops-to-shear-events.patch]: Immutable is not a guarantee. Perhaps document in an own block that its mutability is undefined, so still point towards `#setDrops`?

File: Spigot-API-Patches/0304-Add-drops-to-shear-events.patch [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
++     * Use {@link #setDrops(List)} to change the drops.
```
Comment (Proximyst) [inline Spigot-API-Patches/0304-Add-drops-to-shear-events.patch]:
```suggestion
+     * Get an immutable list of drops for this shearing.
+     * <p>
+     * Use {@link #setDrops(List)} to change the drops.
```

File: Spigot-API-Patches/0304-Add-drops-to-shear-events.patch [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
++        return drops;
++    }
++
++    /**
++     * Sets the drops for the shearing.
++     *
++     * @param drops
++     */
++    public void setDrops(@org.jetbrains.annotations.Nullable java.util.List<org.bukkit.inventory.ItemStack> drops) {
++        this.drops = com.google.common.base.MoreObjects.firstNonNull(drops, com.google.common.collect.Lists.newArrayList());
```
Comment (Proximyst) [inline Spigot-API-Patches/0304-Add-drops-to-shear-events.patch]:
```suggestion
+        this.drops = com.google.common.base.MoreObjects.firstNonNull(drops, java.util.Collections.emptyList());
```

File: Spigot-API-Patches/0304-Add-drops-to-shear-events.patch [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+@@ -17,11 +17,19 @@ public class BlockShearEntityEvent extends BlockEvent implements Cancellable {
+     private final Entity sheared;
+     private final ItemStack tool;
+     private boolean cancelled;
++    private java.util.List<ItemStack> drops; // Paper
+ 
++    @Deprecated // Paper
+     public BlockShearEntityEvent(@NotNull Block dispenser, @NotNull Entity sheared, @NotNull ItemStack tool) {
++        // Paper start
++        this(dispenser, sheared, tool, com.google.common.collect.Lists.newArrayList());
```
Comment (Proximyst) [inline Spigot-API-Patches/0304-Add-drops-to-shear-events.patch]:
```suggestion
+        this(dispenser, sheared, tool, java.util.Collections.emptyList());
```

File: Spigot-API-Patches/0304-Add-drops-to-shear-events.patch [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+@@ -18,12 +18,20 @@ public class PlayerShearEntityEvent extends PlayerEvent implements Cancellable {
+     private final Entity what;
+     private final ItemStack item;
+     private final EquipmentSlot hand;
++    private java.util.List<ItemStack> drops; // Paper
+ 
++    @Deprecated // Paper - include item entity list
+     public PlayerShearEntityEvent(@NotNull Player who, @NotNull Entity what, @NotNull ItemStack item, @NotNull EquipmentSlot hand) {
++        // Paper start
++        this(who, what, item, hand, com.google.common.collect.Lists.newArrayList());
```
Comment (Proximyst) [inline Spigot-API-Patches/0304-Add-drops-to-shear-events.patch]:
```suggestion
+        this(who, what, item, hand, java.util.Collections.emptyList());
```

File: Spigot-API-Patches/0304-Add-drops-to-shear-events.patch [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+ 
+     @Deprecated
+@@ -82,4 +90,25 @@ public class PlayerShearEntityEvent extends PlayerEvent implements Cancellable {
+         return handlers;
+     }
+ 
++    // Paper start
++    /**
++     * Get an immutable list of drops for this shearing.
++     * Use {@link #setDrops(List)} to change the drops.
```
Comment (Proximyst) [inline Spigot-API-Patches/0304-Add-drops-to-shear-events.patch]: See above

File: Spigot-API-Patches/0304-Add-drops-to-shear-events.patch [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
++        return drops;
++    }
++
++    /**
++     * Sets the drops for the shearing.
++     *
++     * @param drops
++     */
++    public void setDrops(@org.jetbrains.annotations.Nullable java.util.List<org.bukkit.inventory.ItemStack> drops) {
++        this.drops = com.google.common.base.MoreObjects.firstNonNull(drops, com.google.common.collect.Lists.newArrayList());
```
Comment (Proximyst) [inline Spigot-API-Patches/0304-Add-drops-to-shear-events.patch]:
```suggestion
+        this.drops = com.google.common.base.MoreObjects.firstNonNull(drops, java.util.Collections.emptyList());
```

File: Spigot-Server-Patches/0736-Add-drops-to-shear-events.patch [hunk 8]
```diff
@@ -0,0 +1,312 @@
...
+ 
+     @Override
+     public void shear(SoundCategory soundcategory) {
++        // Paper start - add drops to shear events
++        this.shear(soundcategory, null);
++    }
++
++    @Override
++    public java.util.List<ItemStack> generateDefaultDrops() {
++        java.util.List<ItemStack> dropEntities = com.google.common.collect.Lists.newArrayList();
```
Comment (Proximyst) [inline Spigot-Server-Patches/0736-Add-drops-to-shear-events.patch]:
```suggestion
+        java.util.List<ItemStack> dropEntities = new java.util.ArrayList<>(5);
```

File: Spigot-Server-Patches/0736-Add-drops-to-shear-events.patch [hunk 9]
```diff
@@ -0,0 +1,312 @@
...
+ 
+ public interface IShearable {
+ 
++    default void shear(SoundCategory soundCategory, @javax.annotation.Nullable java.util.List<org.bukkit.inventory.ItemStack> drops) { this.shear(soundCategory); } // Paper
+     void shear(SoundCategory soundcategory);
+ 
+     boolean canShear();
++    // Paper start - ensure all implementing entities override this
++    default java.util.List<net.minecraft.world.item.ItemStack> generateDefaultDrops() {
++        return com.google.common.collect.Lists.newArrayList();
```
Comment (Proximyst) [inline Spigot-Server-Patches/0736-Add-drops-to-shear-events.patch]:
```suggestion
+        return java.util.Collections.emptyList();
```
Why can't the be empty?

File: Spigot-Server-Patches/0736-Add-drops-to-shear-events.patch [hunk 10]
```diff
@@ -0,0 +1,312 @@
...
+             this.die(); // CraftBukkit - from above
+             // CraftBukkit end
+ 
++            if (drops == null) { // Paper
+             for (int i = 0; i < 5; ++i) {
+                 this.world.addEntity(new EntityItem(this.world, this.locX(), this.e(1.0D), this.locZ(), new ItemStack(this.getVariant().d.getBlock())));
+             }
++            // Paper start - add drops from shear events
++            } else {
++                drops.forEach(bukkitItemStack -> this.dropItemEntity(new EntityItem(this.world, this.locX(), this.e(1.0D), this.locZ(), org.bukkit.craftbukkit.inventory.CraftItemStack.asNMSCopy(bukkitItemStack))));
```
Comment (Proximyst) [inline Spigot-Server-Patches/0736-Add-drops-to-shear-events.patch]: You know the drill (`this.e`).

File: Spigot-Server-Patches/0736-Add-drops-to-shear-events.patch [hunk 11]
```diff
@@ -0,0 +1,312 @@
...
+ 
+     @Override
+     public void shear(SoundCategory soundcategory) {
++        // Paper start
++        this.shear(soundcategory, null);
++    }
++
++    @Override
++    public java.util.List<ItemStack> generateDefaultDrops() {
++        java.util.List<ItemStack> dropEntities = com.google.common.collect.Lists.newArrayList();
```
Comment (Proximyst) [inline Spigot-Server-Patches/0736-Add-drops-to-shear-events.patch]: size

File: Spigot-Server-Patches/0736-Add-drops-to-shear-events.patch [hunk 12]
```diff
@@ -0,0 +1,312 @@
...
+ 
+     @Override
+     public void shear(SoundCategory soundcategory) {
++        // Paper start
++        this.shear(soundcategory, null);
++    }
++
++    @Override
++    public java.util.List<ItemStack> generateDefaultDrops() {
++        return com.google.common.collect.Lists.newArrayList(new ItemStack(Items.dj));
```
Comment (Proximyst) [inline Spigot-Server-Patches/0736-Add-drops-to-shear-events.patch]: OBFHELPER & singletonList

File: Spigot-Server-Patches/0736-Add-drops-to-shear-events.patch [hunk 13]
```diff
@@ -0,0 +1,312 @@
...
++        BlockShearEntityEvent bse = new BlockShearEntityEvent(dispenser, animal.getBukkitEntity(), is, Lists.transform(drops, CraftItemStack::asCraftMirror));
++        // Paper end
+         Bukkit.getPluginManager().callEvent(bse);
+         return bse;
+     }
+ 
++    @Deprecated // Paper - use below
+     public static boolean handlePlayerShearEntityEvent(EntityHuman player, Entity sheared, ItemStack shears, EnumHand hand) {
++        // Paper start
++        PlayerShearEntityEvent event = handlePlayerShearEntityEvent(player, sheared, shears, hand, Lists.newArrayList());
```
Comment (Proximyst) [inline Spigot-Server-Patches/0736-Add-drops-to-shear-events.patch]: emptyList

**Review by Proximyst — CHANGES_REQUESTED** [review]

File: Spigot-API-Patches/0304-Add-drops-to-shear-events.patch (hunk 1 above)
Comment (Machine-Maker) [reply Spigot-API-Patches/0304-Add-drops-to-shear-events.patch]: Is the only reason it’s not guaranteed that someone could construct and fire this event in their own plugin with a mutable list?

Comment (Proximyst) [reply Spigot-API-Patches/0304-Add-drops-to-shear-events.patch]: That and the fact you yourself pass in mutable lists quite a few places in the very same PR.

Comment (Machine-Maker) [reply Spigot-API-Patches/0304-Add-drops-to-shear-events.patch:37]: Isn’t it just Lists.transform? I thought the docs on those said the resulting list is immutable.

**Comment by Machine-Maker** [comment]
Rebased for 1.17.1

**Comment by Machine-Maker** [comment]
Rebased for 1.18.1

File: patches/api/0423-Add-drops-to-shear-events.patch [hunk 14]
```diff
@@ -0,0 +1,111 @@
...
+     private final ItemStack tool;
+     private boolean cancelled;
++    private java.util.List<ItemStack> drops; // Paper
+ 
++    @Deprecated // Paper
+     public BlockShearEntityEvent(@NotNull Block dispenser, @NotNull Entity sheared, @NotNull ItemStack tool) {
++        // Paper start
++        this(dispenser, sheared, tool, java.util.Collections.emptyList());
++    }
++    public BlockShearEntityEvent(@NotNull Block dispenser, @NotNull Entity sheared, @NotNull ItemStack tool, @NotNull java.util.List<ItemStack> drops) {
```
Comment (Sytm) [inline patches/api/0423-Add-drops-to-shear-events.patch]: This constructor is missing the api status annotation the other event has

File: patches/api/0423-Add-drops-to-shear-events.patch [hunk 15]
```diff
@@ -0,0 +1,111 @@
...
+@@ -82,4 +91,26 @@ public class PlayerShearEntityEvent extends PlayerEvent implements Cancellable {
+         return handlers;
+     }
+ 
++    // Paper start - drops API
++    /**
++     * Get the list of drops for this shearing.
++     * <p>
++     * The mutability of this list is undefined so
++     * use {@link #setDrops(java.util.List)} to change the drops.
```
Comment (Lulu13022002) [inline patches/api/0423-Add-drops-to-shear-events.patch]: This list could be unmodifiable to be sure to have at least a consistent behavior in all cases

File: patches/server/0994-Add-drops-to-shear-events.patch [hunk 16]
```diff
@@ -0,0 +1,275 @@
...
+     @Override
+     public void shear(SoundSource shearedSoundCategory) {
++        // Paper start
++        this.shear(shearedSoundCategory, null);
++    }
++
++    @Override
++    public java.util.List<ItemStack> generateDefaultDrops() {
++        int count = this.random.nextInt(3);
++        java.util.List<ItemStack> dropEntities = new java.util.ArrayList<>(count);
```
Comment (Lulu13022002) [inline patches/server/0994-Add-drops-to-shear-events.patch]: The precomputed size should be: 1 + count
or you can increase the count var directly

**Comment by Keviro** [comment]
Any news on this?

**Review by lynxplay — APPROVED** [review]

**Review by Owen1212055 — APPROVED** [review]

**Pull request merged**